### PR TITLE
feat: add shader support by allowing secondary rendering

### PR DESCRIPTION
**Pull request by renancaraujo**
# Description
Allow canvas processing of parts of the fcs tree by enabling the game to be rendered in secondary game widgets (that don't receive events, nor focus nor trigger update cycles, just renders.)

Screenshot 2022-12-05 at 16 25 09

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [ ] I have followed the [Contributor Guide] when preparing my PR.
- [ ] I have updated/added tests for ALL new/updated/fixed functionality.
- [ ] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [ ] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [ ] Yes, this PR is a breaking change.
- [x] No, this PR is not a breaking change.

## Related Issues

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Review by spydon — COMMENTED** [review]
Looks veeery cool (you should add the video to the PR instead)!
A few concerns though:
* Does this have any performance implications for games not using a second games?
* Does this work on the Flutter stable channel?
* Any way we can reduce the code duplications?
* Would it be hard to implement this in the new camera system instead, since this camera system will be going away?

File: packages/flame/lib/src/components/core/component.dart [hunk 1]
```diff
@@ -190,29 +190,38 @@ class Component {
 
   /// Whether the component is currently executing its [onLoad] step.
   bool get isLoading => (_state & _loading) != 0;
+
```
[the remark this answers is not in the document]
Comment (renancaraujo) [reply packages/flame/lib/src/components/core/component.dart:193]: Something wrong is not right, I didn't change these things manually, maybe dartdoc? will fix

File: packages/flame/lib/src/game/game.dart [hunk 2]
```diff
@@ -337,3 +341,253 @@ abstract class Game {
     gameStateListeners.forEach((callback) => callback());
   }
 }
+
+class RenderContext {
```
[the remark this answers is not in the document]
Comment (spydon) [reply packages/flame/lib/src/game/game.dart:345]: Haha was too excited to see that this was a draft :smile:

**Comment by renancaraujo** [comment]
> Does this have any performance implications for games not using a second games?

No, if there's no secondary views, render won't be called 


> Does this work on the Flutter stable channel?

No, The shader api is under constant change. This PR should track changes in the flutter engine API. Until things are stable, I think this pr should be draft 



> Any way we can reduce the code duplications?

Yes, on it


> Would it be hard to implement this in the new camera system instead, since this camera system will be going away?

To not break the current api, the secondary meta info is passed via a pseudo canvas. If you got the canvas, you got the secondary info, if there's any. Bottom line : I don't see why this wouldn't work for the new camera system

**Comment by st-pasha** [comment]
The example looks awesome!

Questions:
1. Does Flutter finally added support for the web platform? I remember this was the reason why we couldn't add the shaders earlier.
2. Is this support going to be in the next Flutter release, do you know what's the timeline here?
3. When I looked at shader API several months ago, it seemed relatively straightforward: create a `ui.FragmentProgram` from GLSL code, then pass that program into the `Paint().shader`, and use that paint to draw whatever -- I presume this structure hasn't changed much?
  - If so, then we shouldn't be needing a new canvas, or a parallel game widget, or anything like that, it should be possible to apply shader through one of our existing methods: a simple paint, a `Decorator`, or a `Layer` (<-- this might be a good opportunity to convert layers into components).
  - for caching purposes we could create a `ShaderCache`, similar to how we have `ImageCache` and `AssetCache`.

**Comment by renancaraujo** [comment]
> The example looks awesome!

I am glad you liked it 

>     1. Does Flutter finally added support for the web platform? I remember this was the reason why we couldn't add the shaders earlier.

Not yet as of yesterday, and that was not the only reason IMO, sampling existing canvas images was very slow (at least 4 frames per pass on my macos when I experimented last time in September). There was a lot of progress with impaller

>     2. Is this support going to be in the next Flutter release, do you know what's the timeline here?

I am not the best person to answer that without guesses. 


>     3. When I looked at shader API several months ago, it seemed relatively straightforward: create a `ui.FragmentProgram` from GLSL code, then pass that program into the `Paint().shader`, and use that paint to draw whatever -- I presume this structure hasn't changed much?


Nope, the changes were made mainly on what kind of code the shaders can include and how uniforms are passed (location annotation is not necessary anymore, for instance).


>     * If so, then we shouldn't need a new canvas, or a parallel game widget, or anything like that, it should be possible to apply shader through one of our existing methods: a simple paint, a `Decorator`, or a `Layer` (<-- this might be an excellent opportunity to convert layers into components).


If one wants to sample a shader with existing textures (like in the case here where we apply the FS on the foreground canvas only), parallel rendering is a must. But that doesn't need to be added to the engine, I extracted  that to a different package just because I can see the usage of this even without fragment shaders


>     * for caching purposes we could create a `ShaderCache`, similar to how we have `ImageCache` and `AssetCache`.

That is a great idea.

**Comment by renancaraujo** [comment]
closing this since I am making a separate package for parallel rendering

**Comment by aytunch** [comment]
@renancaraujo can you please share the link of the repo here once it is public?

**Pull request closed without merging**